### DISCUSSION TITLE
extract keys of type (key type, not its field type)

### DISCUSSION THREAD

**Discussion by regevbr** [millsp/ts-toolbelt#195] [Wiki]
#### Describe the solution you'd like
Since keys of an object can be of type `string|number|symbol` and sometimes types require only string values, it will be handy to have a type that extracts only the string key (or of any type for that matter)

```typescript
type Index = string | number | symbol; // I think we should make that type public!

type KeysOfType<T extends object, M extends Index> = Extract<keyof T, M>;

type t = KeysOfType<{
  a: number;
  2: number;
}, string>; // "a"
```

@pirix-gh thoughts?

**Comment by millsp** [comment]
@regevbr , I currently achieve this by doing `[keys] & [type]`. For example:
```ts
('1' | 1) & string // '1'
```
I rather raise awareness of such techniques instead of writing such types.

So I am marking this as a question :)

**Comment by regevbr** [comment]
That makes sense, although less intuitive in my view. Thought this type will make it more _verbose_ and easier to discover in your library

**Comment by millsp** [comment]
~~You can still be verbose (but more generic) with https://pirix-gh.github.io/ts-toolbelt/modules/_union_filter_.html#filter~~

**Comment by millsp** [comment]
Sorry @regevbr, I meant https://pirix-gh.github.io/ts-toolbelt/modules/_union_select_.html#select

**Comment by regevbr** [comment]
@pirix-gh do you think we can at least expose 
```typescript
type Index = string | number | symbol;
```
in the public interface?

**Comment by millsp** [comment]
@regevbr Yes, I also found myself needing it a few times
